File: native/src/event_loop_group.c
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

#include "crt.h"
#include "exports.h"

#include <aws/io/event_loop.h>

AWS_DOTNET_API
struct aws_event_loop_group *aws_dotnet_event_loop_group_new_default(int num_threads) {
    struct aws_allocator *allocator = aws_dotnet_get_allocator();
    struct aws_event_loop_group *elg = aws_event_loop_group_new_default(allocator, (uint16_t)num_threads, NULL);
    if (elg == NULL) {
        aws_dotnet_throw_exception(aws_last_error(), "Unable to create aws_event_loop_group");
    }

    return elg;
}

AWS_DOTNET_API
void aws_dotnet_event_loop_group_destroy(struct aws_event_loop_group *elg) {
    aws_event_loop_group_release(elg);
}

File: aws-crt/IO/EventLoopGroup.cs
/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */
using System;
using System.Security;
using System.Runtime.InteropServices;

namespace Aws.Crt.IO {
    public class EventLoopGroup {

        [SecuritySafeCritical]
        internal static class API
        {
            public delegate Handle aws_dotnet_event_loop_group_new_default(int numThreads);
            public delegate void aws_dotnet_event_loop_group_destroy(IntPtr elg);

            public static aws_dotnet_event_loop_group_new_default make_new_default = NativeAPI.Bind<aws_dotnet_event_loop_group_new_default>();
            public static aws_dotnet_event_loop_group_destroy destroy = NativeAPI.Bind<aws_dotnet_event_loop_group_destroy>();
        }

        internal class Handle : CRT.Handle
        {
            protected override bool ReleaseHandle() {
                API.destroy(handle);
                return true;
            }
        }
        
        internal Handle NativeHandle { get; private set; }

        public EventLoopGroup(int numThreads=1) {
            NativeHandle = API.make_new_default(numThreads);
        }
    }
 }
 